Move needs_dynamic_casting to a non-CUDA specific file.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38817 Add dynamic_cast asserts to CPU Loops.
* #38815 Support void return type in TensorIteratorDynamicCasting checks.
* **#38813 Move needs_dynamic_casting to a non-CUDA specific file.**
* #38810 Improve error checking of CUDALoops.
* #38809 Improve cpu/Loops.h arity asserts.

We are going to apply this check to CPU (with some changes), so just moving this in preparation.

The code is just cut-pasted here, no behavioral change.

Differential Revision: [D21670554](https://our.internmc.facebook.com/intern/diff/D21670554)

gchanan committed:
Move needs_dynamic_casting to a non-CUDA specific file.

We are going to apply this check to CPU (with some changes), so just moving this in preparation.

The code is just cut-pasted here, no behavioral change.

[ghstack-poisoned]

zasdfgbnm commented on aten/src/ATen/native/TensorIteratorDynamicCasting.h:
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <ATen/detail/FunctionTraits.h>
```
maybe also
```
#include <ATen/native/TensorIterator.h>
#include <c10/core/ScalarType.h>
```

gchanan replied:
good call.

gchanan committed:
Update on "Move needs_dynamic_casting to a non-CUDA specific file."


We are going to apply this check to CPU (with some changes), so just moving this in preparation.

The code is just cut-pasted here, no behavioral change.

Differential Revision: [D21670554](https://our.internmc.facebook.com/intern/diff/D21670554)

[ghstack-poisoned]

gchanan committed:
Update on "Move needs_dynamic_casting to a non-CUDA specific file."


We are going to apply this check to CPU (with some changes), so just moving this in preparation.

The code is just cut-pasted here, no behavioral change.

Differential Revision: [D21670554](https://our.internmc.facebook.com/intern/diff/D21670554)

[ghstack-poisoned]

gchanan committed:
Update on "Move needs_dynamic_casting to a non-CUDA specific file."


We are going to apply this check to CPU (with some changes), so just moving this in preparation.

The code is just cut-pasted here, no behavioral change.

Differential Revision: [D21670554](https://our.internmc.facebook.com/intern/diff/D21670554)

[ghstack-poisoned]
